File: client/handlers/cloud_client.py
# handlers/cloud_client.py
import time
import requests
from typing import Dict, Any

from protocol.packet import PacketManager

class CloudClient:
    """Клиент для облачного взаимодействия"""
    
    def __init__(self, api_token: str, device_token: str, device_id: str = "python_client", server_url: str = "https://wakelink.deadboizxc.org"):
        self.api_token = api_token
        self.device_token = device_token
        self.device_id = device_id
        self.server_url = server_url
        self.packet_manager = PacketManager(device_token, device_id)
        print(f"☁️ Cloud Client: {server_url} (Device: {device_id})")

    def send_command(self, command: str, data: Dict = None) -> Dict[str, Any]:
        """Отправка команды через облако"""
        encrypted_command = self.packet_manager.create_encrypted_command(command, data)
        
        # СТРУКТУРА КАК НА СЕРВЕРЕ: только device_token
        push_payload = {
            "device_token": self.device_token,
            "msg_type": "command",
            "encrypted_payload": encrypted_command,
            "is_response": False
        }

        try:
            push_response = requests.post(
                f"{self.server_url}/api/push",
                json=push_payload,
                timeout=10
            )
            
            if push_response.status_code != 200:
                return {"status": "error", "error": f"Push failed: {push_response.status_code}"}
            
            # Ждем ответ
            time.sleep(2)
            
            # СТРУКТУРА КАК НА СЕРВЕРЕ: device_token и device_id
            pull_payload = {
                "device_token": self.device_token,
                "device_id": self.device_id
            }
            
            pull_response = requests.post(
                f"{self.server_url}/api/pull",
                json=pull_payload,
                timeout=10
            )
            
            if pull_response.status_code != 200:
                return {"status": "error", "error": f"Pull failed: {pull_response.status_code}"}
            
            pull_data = pull_response.json()
            
            # Обрабатываем сообщения
            if "messages" in pull_data and pull_data["messages"]:
                for msg in pull_data["messages"]:
                    if msg.get("direction") == "to_device":
                        encrypted_response = msg.get("data", "")
                        if encrypted_response:
                            return self.packet_manager.decrypt_response(encrypted_response)
            
            return {"status": "timeout", "message": "No response from device"}
                    
        except Exception as e:
            return {"status": "error", "error": str(e)}

File: firmware/WakeLink/cloud_client.cpp
// cloud_client.cpp
#include "cloud_client.h"
#include "command.h"
#include "packet.h"

extern PacketManager packetManager;
extern String API_TOKEN;
extern String DEVICE_ID;

// =============================
// 1. PUSH
void pushCloud(const String& msg_type, const String& encrypted_payload, bool is_response) {
    if (!strlen(cfg.server_url)) return;

    HTTPClient http;
    String url = String(cfg.server_url) + "/api/push";
    http.begin(clientSecure, url);
    http.addHeader("Content-Type", "application/json");

    JsonDocument req;
    req["device_token"] = API_TOKEN;
    req["msg_type"] = msg_type;
    req["encrypted_payload"] = encrypted_payload;
    req["is_response"] = is_response;

    String payload;
    serializeJson(req, payload);
    int code = http.POST(payload);
    http.end();

    if (code == 200) {
        Serial.printf("[CLOUD] Pushed %s: %d bytes\n", msg_type.c_str(), encrypted_payload.length());
    } else {
        Serial.printf("[CLOUD] Push failed: %d\n", code);
    }
}

// =============================
// 2. PULL
void pullCloud() {
    if (!strlen(cfg.server_url)) return;

    HTTPClient http;
    String url = String(cfg.server_url) + "/api/pull";
    http.begin(clientSecure, url);
    http.addHeader("Content-Type", "application/json");

    JsonDocument req;
    req["device_token"] = API_TOKEN;
    req["device_id"] = DEVICE_ID;

    String payload;
    serializeJson(req, payload);
    int code = http.POST(payload);

    if (code != 200) {
        Serial.printf("[CLOUD] Pull failed: %d\n", code);
        http.end();
        return;
    }

    String resp = http.getString();
    http.end();

    JsonDocument doc;
    DeserializationError err = deserializeJson(doc, resp);
    if (err) {
        Serial.printf("[CLOUD] JSON error: %s\n", err.c_str());
        return;
    }

    JsonArray msgs = doc["messages"];
    if (msgs.isNull()) return;

    for (JsonObject m : msgs) {
        String type = m["type"] | "";
        String encrypted = m["data"] | "";
        String direction = m["direction"] | "";

        if (encrypted.length() == 0) continue;
        if (direction != "to_device") continue;

        JsonDocument incoming = packetManager.decryptIncomingPacket(encrypted);
        if (incoming["status"] != "success") {
            const char* err = incoming["error"] | "DECRYPT_FAILED";
            JsonDocument error;
            error["status"] = "error";
            error["error"] = err;
            String encryptedResp = packetManager.createEncryptedResponse(error);
            pushCloud("response", encryptedResp, true);
            continue;
        }

        const char* cmd = incoming["command"];
        if (!cmd || strlen(cmd) == 0) {
            JsonDocument error;
            error["status"] = "error";
            error["error"] = "NO_COMMAND";
            String encryptedResp = packetManager.createEncryptedResponse(error);
            pushCloud("response", encryptedResp, true);
            continue;
        }

        JsonObject data = incoming["data"].as<JsonObject>();  // ← Прямой доступ
        JsonDocument result = CommandManager::executeCommand(String(cmd), data);
        String encryptedResp = packetManager.createEncryptedResponse(result);
        pushCloud("response", encryptedResp, true);
    }
}

// =============================
// УТИЛИТЫ
void handleCloud() {
    static unsigned long lastPull = 0;
    if (millis() - lastPull >= 5000) {
        pullCloud();
        lastPull = millis();
    }
}

void initCloud() {
    Serial.println("[CLOUD] Cloud client initialized");
}